torch/csrc/jit/codegen/cuda/lower_allocation.cpp

```diff
@@ -145,7 +146,146 @@ class AllocationInserter : public kir::MutableIrVisitor {
           id->iterType() == IterType::BroadcastWithoutStride) {
         continue;
       }
-      alloc_dims.push_back(id->extent());
+      auto extent = id->extent();
+      // Use halo-extended extent if found
+      auto halo_extent = gpu_lower->haloInfo().getRootAxisInfo(id);
+      if (halo_extent.width() != 0) {
+        extent = ir_builder.addExpr(
+            extent, ir_builder.create<kir::Int>(halo_extent.width()));
+      }
+      alloc_dims.push_back(extent);
+    }
+
+    return alloc_dims;
+  }
+
+  // Get allocation extents of root axes with halo
+  //
+  // Allocation can be done with leaf IDs with halo as well, but
+  // allocation size could be larger than necessary.
+  //
+  // For example, suppose the shift offset of an axis is 1. When it is
+  // split by N, the halo size of the inner output is N+1. When the
+  // allocation only has the inner split output, the allocation size
+  // would be N+1. Suppose that ID is further split by M, the output
+  // extents would be N/M and M+1. The allocation size based on the
+  // leaves would be N/M*(M+1) or N+N/M, which is larger than N+1.
+  //
+  // This function tries to propagate back halo informatin to root
+  // axes to avoid inflating allocations. It fails when merged domains
+  // are split and only one of the split outputs is used for
+  // allocations since in such a case we can't un-merge and properly
+  // determine the extents of the merge inputs. If that would result
```

Comment:
I think you're trying to say here it does result in an exception.

Reply:
Thanks. Yes. Fixed.